src/basilisp/contrib/nrepl_server.lpy

```diff
@@ -0,0 +1,479 @@
+;; adapted from
+;; https://github.com/babashka/nbb/blob/bca8b5017a06768eb35d02a2d6233ca9c6c2f692/src/nbb/impl/nrepl_server.cljs
+(ns basilisp.contrib.nrepl-server
+  "A port of `nbb <https://github.com/babashka/nbb>`_'s nREPL server implementation to basilisp."
+  (:require [basilisp.contrib.bencode :as bc]
+            [basilisp.string :as str])
+  (:import logging
+           [basilisp.lang.runtime :as rt]
+           socketserver
+           sys
+           traceback
+           uuid))
+
+(def ^:private logger
+  "The logger for this namespace."
+  (logging/getLogger (namespace ::)))
+
+(defmacro ^:private debug [& values]
+  (when (.isEnabledFor logger logging/DEBUG)
+    `(.debug logger (str/join " " [~@values]))))
+(defmacro ^:private info [& values]
+  (when (.isEnabledFor logger logging/INFO)
+    `(.info logger (str/join " " [~@values]))))
+(defmacro ^:private warn [& values]
+  (when (.isEnabledFor logger logging/WARNING))
+  `(.warning logger (str/join " " [~@values])))
+(defmacro ^:private error [& values]
+  ;; we want error to always output to `sys/stderr` unfiltered.
+  `(.write sys/stderr (str ~(keyword (str *ns*) "error") " " (str/join " " [~@values]) \newline)))
+
+(definterface ^:private IReadWrite
+  ;; Pythonic interface for creating `sys/stdout` like File objects.
+  (read [])
+  (write [value]))
+
+(deftype ^:private StreamOutFn [out-fn]
+  ;; A type to use as replacement binding for writing to `sys/stdout`
+  ;; stream, so that the output ``value`` is passed to ``out-fn``
+  ;; instead.
+  IReadWrite
+  (read [_this]
+    (throw (python/RuntimeError (str ::logic-error " shouldn't be possible to get here..."))))
```

Comment:
FWIW I typically use `(ex-info ...)` as in Clojure for cases where no more specific options exist.

Reply:
I've realised the sys/stdout "API" does not require a read operation, so I've removed it and replaced it with a no-op `flush` method, and renamed it to `IStdOut` accordingly.